src/python/pants/bin/daemon_pants_runner.py

```diff
@@ -12,36 +12,50 @@
 from builtins import open, zip
 from contextlib import contextmanager
 
-from future.utils import PY3, raise_with_traceback
-from setproctitle import setproctitle as set_process_title
-
 from pants.base.build_environment import get_buildroot
 from pants.base.exception_sink import ExceptionSink, SignalHandler
 from pants.base.exiter import PANTS_FAILED_EXIT_CODE, PANTS_SUCCEEDED_EXIT_CODE, Exiter
 from pants.bin.local_pants_runner import LocalPantsRunner
+from pants.init.logging import encapsulated_global_logger
 from pants.init.util import clean_global_runtime_state
 from pants.java.nailgun_io import NailgunStreamStdinReader, NailgunStreamWriter
-from pants.java.nailgun_protocol import ChunkType, NailgunProtocol
-from pants.pantsd.process_manager import ProcessManager
+from pants.java.nailgun_protocol import ChunkType, MaybeShutdownSocket, NailgunProtocol
 from pants.util.contextutil import hermetic_environment_as, stdio_as
 from pants.util.socket import teardown_socket
 
 
 class DaemonSignalHandler(SignalHandler):
 
   def handle_sigint(self, signum, _frame):
+    self.daemon.shutdown()
     raise KeyboardInterrupt('remote client sent control-c!')
 
+  def handle_sigterm(self, signum, _frame):
+    try:
+      self.daemon.shutdown()
+    except Exception:
+      pass
+
+
+class NoopExiter(Exiter):
+  def exit(self, result, *args, **kwargs):
+    if result != 0:
+      # TODO this exception was not designed for this, but it happens to do what we want.
```

Comment:
Expanding this comment would be good: what is it that we want from the `NoopExiter`? As with many things, I expect that we can clean it up post merge, but would be good to include a bit more context in case ... that takes a while.